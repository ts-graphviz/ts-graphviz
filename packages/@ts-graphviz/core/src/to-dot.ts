import { DotObjectModel } from '@ts-graphviz/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '@ts-graphviz/ast';

/**
 * This interface provides options for converting a model to DOT.
 * @group Convert Model to DOT
 */
export interface ToDotOptions {
  /**
   * Options for converting the model to DOT.
   */
  convert?: ConvertFromModelOptions;
  /**
   * Options for printing DOT.
   */
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Subgraph}, {@link Node}, and {@link Edge}
 * @param options Optional options for the conversion.
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
