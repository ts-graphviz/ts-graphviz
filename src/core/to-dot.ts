import { DotObjectModel } from '#lib/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '#lib/ast';

/**
 * @group Convert Model to DOT
 * @alpha
 */
export interface ToDotOptions {
  convert?: ConvertFromModelOptions;
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Node}, and {@link Edge}
 * @param options
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
