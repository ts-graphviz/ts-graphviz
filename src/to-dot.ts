import { DotObjectModel } from './common/index.js';
import { stringify, fromModel, PrintOptions } from './ast/index.js';
import { ConvertOptions } from './ast/index.js';

export interface ToDotOptions {
  convert?: ConvertOptions;
  print?: PrintOptions;
}

export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
