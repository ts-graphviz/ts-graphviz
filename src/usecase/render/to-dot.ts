import { Renderer, Dot } from './renderer.js';

export function toDot(object: Dot): string {
  const renderer = new Renderer();
  return renderer.render(object);
}
