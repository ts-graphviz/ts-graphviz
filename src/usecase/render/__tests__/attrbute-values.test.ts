import { toDot } from '../to-dot.js';

describe('AttributesValue rendering', () => {
  const DoubleQuotedValuePattern = /^".+"$/ms;
  const HTMLLikeValuePattern = /^<.+>$/ms;

  describe('Literal', () => {
    describe('plain string', () => {
      test.each([
        ['te"st'],
        [
          `
          test
          `,
        ],
      ])('toDot result should be quoted', (value) => {
        const dot = toDot(value);
        expect(dot).toMatch(DoubleQuotedValuePattern);
        expect(dot).not.toMatch(HTMLLikeValuePattern);
        expect(dot).toMatchSnapshot();
      });
    });

    describe('escape characters', () => {
      test.each([
        ['colon', '1:2', '"1:2"'],
        ['semi-colon', '1;2', '"1;2"'],
        ['equal', '1=2', '"1=2"'],
        ['dash', '1-2', '"1-2"'],
        ['sharp', '1#2', '"1#2"'],
        ['space', '1 2', '"1 2"'],
        ['newline', '1\n2', '"1\\n2"'],
        ['tab', '1\t2', '"1\t2"'],
        ['double-quotation', '1"2', '"1\\"2"'],
        ['slash', '1/2', '"1/2"'],
      ])('escaped if literal contains %s character', (_name, value, expected) => {
        const dot = toDot(value);
        expect(dot).toBe(expected);
      });
    });

    describe('HTML Like string', () => {
      test.each([
        ['<<B>hoge</B>>'],
        [
          `<
          <U>hoge</U>
          >`,
        ],
        [
          `
          <<I>hoge</I>
          >
          `,
        ],
      ])('toDot result should not be quoted', (value) => {
        const dot = toDot(value);
        expect(dot).not.toMatch(DoubleQuotedValuePattern);
        expect(dot).toMatch(HTMLLikeValuePattern);
        expect(dot).toMatchSnapshot();
      });
    });
  });
});
