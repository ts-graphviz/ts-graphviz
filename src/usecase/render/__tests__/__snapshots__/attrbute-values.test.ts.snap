// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributesValue rendering Literal HTML Like string toDot result should not be quoted 1`] = `"<<B>hoge</B>>"`;

exports[`AttributesValue rendering Literal HTML Like string toDot result should not be quoted 2`] = `
"<
          <U>hoge</U>
          >"
`;

exports[`AttributesValue rendering Literal HTML Like string toDot result should not be quoted 3`] = `
"<<I>hoge</I>
          >"
`;

exports[`AttributesValue rendering Literal plain string toDot result should be quoted 1`] = `"\\"te\\\\\\"st\\""`;

exports[`AttributesValue rendering Literal plain string toDot result should be quoted 2`] = `"\\"\\\\n          test\\\\n          \\""`;
